jsparagus/gen.py

```diff
@@ -2969,6 +2985,99 @@ def remove_all_unreachable_state(self, verbose, progress):
         for s in self.states:
             s.rewrite_state_indexes(state_map)
 
+    def fold_reduce_cascade(self, verbose, progress):
+        # Reduce actions, especially with lookahead tokens, have a tendency of
+        # reducing more than one non-terminal in cascade as it reduces multiple
+        # productions. This function is here to avoid these intermediate reduce
+        # actions and only keep the last one with all the intermediate actions
+        # squashed in the same sequence.
+        def collect_common_history(history, aps_list):
+            "Collect the list of futures which are accepting the same terms."
+
+            # Recall the history to be returned, if it reduces at least the
+            # same tokens as the initial reduce action. Otherwise, we might be
+            # visiting reduce actions which are reducing the lookahead tokens,
+            # which is not yet handled by the backend, as we have no shift
+            # actions to emulate what the shifting would have accomplished.
```

Comment:
This comment is unclear to me, even though I think I do know roughly what this function is trying to accomplish. But what is the contract? What are the arguments? What is the return value?